File: CPP/ByRound/2096/2096_E_Wonderful_Teddy_Bears.cpp
// Problem: CF 2096 E - Wonderful Teddy Bears
// https://codeforces.com/contest/2096/problem/E

/*
E. Wonderful Teddy Bears

Algorithm/Technique: Greedy + Simulation
Time Complexity: O(n) per test case
Space Complexity: O(1)

This problem asks for the minimum number of operations to make a sequence of black (B) and pink (P) teddy bears
beautiful — i.e., all B's come before all P's. Each operation allows reordering three consecutive elements so that
all B's are to the left of all P's in that group.

The approach uses a greedy idea:
- Count how many B's we've seen so far and track how many of those B’s are at odd positions (1-indexed).
- For each B encountered, compute how much "disorder" it contributes.
- At the end, the final answer combines this disorder with a correction term based on balancing B's between even/odd
  positions to ensure minimal operations.

Key Observations:
- We process the string from left to right.
- For every 'B', we consider its effect on the total cost:
    - It increases `a` (count of B's so far).
    - If its index is even, it contributes to `b`.
    - Add `i + 1 - a` to accumulate the cost from previous misplacements.
- A correction term uses absolute difference between half the total count of B's and count of B’s in even positions.

*/

#include <bits/std_abs.h>
#include <iostream>
#include <string>

using namespace std;
using i64 = long long;

int T;
int n;
i64 ans, a, b;
string s;

void solve() {
  a = b = ans = 0; // reset counters
  cin >> n >> s;
  for (int i = 0; i < n; ++i) {
    if (s[i] == 'B') { // when we see a black bear
      a++; // increment count of black bears seen so far
      b += !((i + 1) & 1); // increment if this black bear is at an even index (1-based)
      ans += i + 1 - a; // accumulate cost due to disorder introduced by this B
    }
  }
  cout << (ans + abs(a / 2 - b)) / 2 << '\n'; // compute result using derived formula
}

int main() {
  ios::sync_with_stdio(false);
  cin.tie(0);
  cout.tie(0);
  cin >> T;
  while (T--)
    solve();
  return 0;
}


// https://github.com/VaHiX/codeForces/

File: Python/ByTier/E/2096_E_Wonderful_Teddy_Bears.py
# Problem: CF 2096 E - Wonderful Teddy Bears
# https://codeforces.com/contest/2096/problem/E

"""
E. Wonderful Teddy Bears

Purpose:
This code solves the problem of reordering teddy bears to make a "beautiful" arrangement,
where all black bears (B) are to the left of all pink bears (P). The allowed operation is
to select three consecutive positions and sort them so B's are to the left of P's.

Algorithm:
- The approach is based on tracking how many B's have been placed correctly.
- For each B encountered in the string, we determine if it is already in a correct position,
  or needs to be moved via an operation.
- We maintain two counters: 'ev' for even-indexed B placements (used when placing B's),
  and 'od' for odd-indexed B placements (used to track alternative placements).
- For each B, we calculate how many operations are needed based on the current index and
  expected placement.

Time Complexity: O(n), where n is the length of the string s.
Space Complexity: O(1), constant space usage.

"""

import sys

input = sys.stdin.readline
for _ in range(int(input())):
    n = int(input())
    s = input().strip()
    num = 0
    sol = 0
    ev = 0  # Tracks even position for B placement
    od = 1  # Tracks odd position for B placement
    cnt = s.count("B")  # Total number of black bears
    
    for i in range(n):
        if s[i] == "B":
            # Determine whether to use even or odd tracking strategy
            if (ev < cnt and (i - ev) % 2 == 0 and i >= ev) or od >= cnt or i < od:
                # Use even strategy to assign this B
                sol += (i - ev + 1) // 2
                ev += 2
            else:
                # Use odd strategy to assign this B
                sol += (i - od + 1) // 2
                od += 2
    print(sol)


# https://github.com/VaHiX/codeForces/